keyring does not compile on MacOS 15.3.2 with GCC 14

Hello there,

After upgrading to R 4.5.0 I've had the pelasure of reinstalling all my libraries, and found that `keyring` would fail compilation. I am running an M3 Mac with GCC 14 installed using Homebrew, because Apple `clang` is always quite outdated. I am therefore forcing `CC=gcc-14` and `CXX=g++-14` in my ~/.R/Makevars for ease of use.

`install.packages("keyring")` fails with the following:
```
> install.packages("keyring")
Installing package into ‘/opt/homebrew/lib/R/4.5/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://cran.rstudio.com/src/contrib/keyring_1.3.2.tar.gz'
Content type 'application/x-gzip' length 46498 bytes (45 KB)
==================================================
downloaded 45 KB

* installing *source* package ‘keyring’ ...
** this is package ‘keyring’ version ‘1.3.2’
** package ‘keyring’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
using C compiler: ‘gcc-14 (Homebrew GCC 14.2.0_1) 14.2.0’
using SDK: ‘MacOSX15.2.sdk’
gcc-14 -I"/opt/homebrew/Cellar/r/4.5.0/lib/R/include" -DNDEBUG   -I/opt/homebrew/opt/gettext/include -I/opt/homebrew/opt/readline/include -I/opt/homebrew/opt/xz/include -I/opt/homebrew/include    -fPIC  -g -O2  -c init.c -o init.o
gcc-14 -I"/opt/homebrew/Cellar/r/4.5.0/lib/R/include" -DNDEBUG   -I/opt/homebrew/opt/gettext/include -I/opt/homebrew/opt/readline/include -I/opt/homebrew/opt/xz/include -I/opt/homebrew/include    -fPIC  -g -O2  -c keyring_macos.c -o keyring_macos.o
keyring_macos.c: In function 'keyring_macos_create':
keyring_macos.c:313:10: error: expected expression before '#pragma'
  313 | # pragma GCC diagnostic pop
      |          ^~~
keyring_macos.c:312:21: error: too few arguments to function 'SecKeychainCreate'
  312 |   OSStatus status = SecKeychainCreate(
      |                     ^~~~~~~~~~~~~~~~~
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks/Security.framework/Headers/SecImportExport.h:41,
                 from /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks/Security.framework/Headers/Security.h:35,
                 from keyring_macos.c:5:
/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks/Security.framework/Headers/SecKeychain.h:309:10: note: declared here
  309 | OSStatus SecKeychainCreate(const char *pathName, UInt32 passwordLength, const void * __nullable password, Boolean promptUser, SecAccessRef __nullable initialAccess, SecKeychainRef * __nonnull CF_RETURNS_RETAINED keychain)
      |          ^~~~~~~~~~~~~~~~~
make: *** [keyring_macos.o] Error 1
ERROR: compilation failed for package ‘keyring’
* removing ‘/opt/homebrew/lib/R/4.5/site-library/keyring’
* restoring previous ‘/opt/homebrew/lib/R/4.5/site-library/keyring’
Warning in install.packages :
  installation of package ‘keyring’ had non-zero exit status

The downloaded source packages are in
	‘/private/var/folders/j3/kp576yhx6hg05g7g2cs03y_84f_rmw/T/Rtmp4ojb50/downloaded_packages’
```

This seems to be because the `pop` should rather happen after the block [here](https://github.com/r-lib/keyring/blob/369083ac29468c15d1d0ef3d6a790df47a6cd7b8/src/keyring_macos.c#L338). 

Reverting to `clang` / `clang++` solves the issue, although my understanding is the pragma is not properly placed and this should be addressed.

Happy to provide more debug info if needed
Thomas

--- a/src/keyring_macos.c
+++ b/src/keyring_macos.c
@@ -335,12 +335,12 @@ SEXP keyring_macos_create(SEXP keyring, SEXP password) {
 # pragma GCC diagnostic push
 # pragma GCC diagnostic ignored "-Wdeprecated-declarations"
   OSStatus status = SecKeychainCreate(
-# pragma GCC diagnostic pop
     ckeyring,
     /* passwordLength = */ (UInt32) strlen(cpassword),
     (const void*) cpassword,
     /* promptUser = */ 0, /* initialAccess = */ NULL,
     &result);
+# pragma GCC diagnostic pop
 
   keyring_macos_handle_status("cannot create keychain", status);
 

